File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/EndTransactionLogRecord.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Diagnostics.CodeAnalysis;
    using System.Fabric;
    using System.Fabric.Common;
    using System.IO;

    internal sealed class EndTransactionLogRecord : TransactionLogRecord
    {
        private const uint ApproximateDiskSpaceUsed = sizeof(bool);

        private const int SizeOnWireIncrement = sizeof(int) + sizeof(bool);

        private readonly bool isThisReplicaTransaction;

        // This is the only field persisted
        private bool isCommitted;

        private ArraySegment<byte> replicatedData;

        [SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1126:PrefixCallsCorrectly", Justification = "Reviewed.")]
        internal EndTransactionLogRecord(LogRecordType recordType, ulong recordPosition, long lsn)
            : base(recordType, recordPosition, lsn)
        {
            Utility.Assert(
                recordType == LogRecordType.EndTransaction,
                "LogRecordType {0} != EndTransaction",
                recordType);
            this.ChildTransactionRecord = null;
            this.isCommitted = false;
            this.isThisReplicaTransaction = false;
        }

        [SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1126:PrefixCallsCorrectly", Justification = "Reviewed.")]
        internal EndTransactionLogRecord(TransactionBase transaction, bool isCommitted, bool isThisReplicaTransaction)
            : base(LogRecordType.EndTransaction, transaction, InvalidTransactionLogRecord)
        {
            Utility.Assert(
                (isThisReplicaTransaction == true) || (isCommitted == false),
                "This should be a replica transaction or it should not be committed. IsThisReplicaTransaction : {0}, IsCommited: {1} ",
                isThisReplicaTransaction, isCommitted);
            this.ChildTransactionRecord = null;
            this.isCommitted = isCommitted;
            this.isThisReplicaTransaction = isThisReplicaTransaction;
            this.UpdateApproximateDiskSize();
        }

        internal bool IsCommitted
        {
            get { return this.isCommitted; }

            set { this.isCommitted = value; }
        }

        internal override bool IsLatencySensitiveRecord
        {
            get { return true; }
        }

        internal bool IsThisReplicaTransaction
        {
            get { return this.isThisReplicaTransaction; }
        }

        protected override int GetSizeOnWire()
        {
            return base.GetSizeOnWire() + SizeOnWireIncrement;
        }

        internal override void InvalidateReplicatedDataBuffers(SynchronizedBufferPool<BinaryWriter> pool)
        {
            base.InvalidateReplicatedDataBuffers(pool);
            this.replicatedData = new ArraySegment<byte>();
        }

        protected override void Read(BinaryReader br, bool isPhysicalRead)
        {
            base.Read(br, isPhysicalRead);
            this.ReadPrivate(br, isPhysicalRead);
            return;
        }

        protected override void ReadLogical(OperationData operationData, ref int index)
        {
            base.ReadLogical(operationData, ref index);

            using (var br = new BinaryReader(IncrementIndexAndGetMemoryStreamAt(operationData, ref index)))
            {
                this.ReadPrivate(br, false);
            }
        }

        protected override void Write(BinaryWriter bw, OperationData operationData, bool isPhysicalWrite, bool forceRecomputeOffsets)
        {
            base.Write(bw, operationData, isPhysicalWrite, forceRecomputeOffsets);

            if (this.replicatedData.Array == null)
            {
                var startingPos = bw.BaseStream.Position;
                bw.BaseStream.Position += sizeof(int);

                bw.Write(this.isCommitted);

                var endPosition = bw.BaseStream.Position;
                var sizeOfSection = checked((int) (endPosition - startingPos));
                bw.BaseStream.Position = startingPos;
                bw.Write(sizeOfSection);
                bw.BaseStream.Position = endPosition;

                this.replicatedData = CreateArraySegment(startingPos, bw);
            }

            operationData.Add(this.replicatedData);
        }

        private void ReadPrivate(BinaryReader br, bool isPhysicalRead)
        {
            var startingPosition = br.BaseStream.Position;
            var sizeOfSection = br.ReadInt32();
            var endPosition = startingPosition + sizeOfSection;

            this.isCommitted = br.ReadBoolean();

            // Jump to the end of the section ignoring fields that are not understood.
            Utility.Assert(endPosition >= br.BaseStream.Position, "Could not have read more than section size.");
            br.BaseStream.Position = endPosition;

            this.UpdateApproximateDiskSize();
        }

        private void UpdateApproximateDiskSize()
        {
            this.ApproximateSizeOnDisk += ApproximateDiskSpaceUsed;
        }
    }
}

File: src/prod/src/data/txnreplicator/logrecordlib/EndTransactionLogRecord.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

const ULONG EndTransactionLogRecord::DiskSpaceUsed = sizeof(bool) + sizeof(LONG32);
const ULONG EndTransactionLogRecord::SizeOnWireIncrement = sizeof(LONG32) + sizeof(bool);

EndTransactionLogRecord::EndTransactionLogRecord(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog)
    : TransactionLogRecord(recordType, recordPosition, lsn, invalidPhysicalLogRecord, invalidTransactionLog)
    , isCommitted_(false)
    , isThisReplicaTransaction_(false)
{
    ASSERT_IFNOT(recordType == LogRecordType::Enum::EndTransaction, "Expected end xact record type, actual: {0}", recordType);
}

EndTransactionLogRecord::EndTransactionLogRecord(
    __in TransactionBase & transaction,
    __in bool isCommitted,
    __in bool isThisReplicaTransaction,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog)
    : TransactionLogRecord(LogRecordType::EndTransaction, transaction, invalidPhysicalLogRecord, invalidTransactionLog, &invalidTransactionLog)
    , isCommitted_(isCommitted)
    , isThisReplicaTransaction_(isThisReplicaTransaction)
{
    UpdateApproximateDiskSize();
}

EndTransactionLogRecord::~EndTransactionLogRecord()
{
}

EndTransactionLogRecord::SPtr EndTransactionLogRecord::Create(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog,
    __in KAllocator & allocator)
{
    EndTransactionLogRecord* pointer = _new(ENDTXLOGRECORD_TAG, allocator) EndTransactionLogRecord(
        recordType, 
        recordPosition, 
        lsn, 
        invalidPhysicalLogRecord, 
        invalidTransactionLog);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return EndTransactionLogRecord::SPtr(pointer);
}

EndTransactionLogRecord::SPtr EndTransactionLogRecord::Create(
    __in TransactionBase & transaction,
    __in bool isCommitted,
    __in bool isThisReplicaTransaction,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog,
    __in KAllocator & allocator)
{
    EndTransactionLogRecord* pointer = _new(ENDTXLOGRECORD_TAG, allocator) EndTransactionLogRecord(
        transaction,
        isCommitted,
        isThisReplicaTransaction,
        invalidPhysicalLogRecord,
        invalidTransactionLog);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return EndTransactionLogRecord::SPtr(pointer);;
}

std::wstring EndTransactionLogRecord::ToString() const
{
    wstring boolean = isCommitted_ ? L"true" : L"false"; 

    std::wstring logRecordString = Constants::CompEndlJSON;

    logRecordString += L"Committed" + Constants::DivisionBoolJSON + boolean;
    logRecordString += Constants::CloseJSON;

    std::wstring logRecordWstring(logRecordString.begin(), logRecordString.end());

    return __super::ToString() + logRecordString;
}

void EndTransactionLogRecord::UpdateApproximateDiskSize()
{
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + DiskSpaceUsed;
}

void EndTransactionLogRecord::Read(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    __super::Read(binaryReader, isPhysicalRead);
    ReadPrivate(binaryReader, isPhysicalRead);
}

void EndTransactionLogRecord::ReadLogical(
    __in OperationData const & operationData,
    __inout INT & index)
{
    __super::ReadLogical(operationData, index);
    KBuffer::CSPtr buffer = OperationData::IncrementIndexAndGetBuffer(operationData, index);
    BinaryReader reader(*buffer, GetThisAllocator());
    ReadPrivate(reader, false);
}

void EndTransactionLogRecord::ReadPrivate(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    UNREFERENCED_PARAMETER(isPhysicalRead);

    // Read Metadata size
    ULONG32 startingPosition = binaryReader.Position;
    ULONG32 sizeOfSection = 0;
    binaryReader.Read(sizeOfSection);
    ULONG32 endPosition = startingPosition + sizeOfSection;

    binaryReader.Read(isCommitted_);

    ASSERT_IFNOT(
        endPosition >= binaryReader.Position,
        "Invalid end xact op log record, end position: {0} reader position:{1}",
        endPosition, binaryReader.Position);

    binaryReader.Position = endPosition;

    UpdateApproximateDiskSize();
}

void EndTransactionLogRecord::Write(
    __in BinaryWriter & binaryWriter,
    __inout OperationData & operationData,
    __in bool isPhysicalWrite,
    __in bool forceRecomputeOffsets)
{
    __super::Write(binaryWriter, operationData, isPhysicalWrite, forceRecomputeOffsets);

    if (replicatedData_ == nullptr)
    {        
        ULONG32 startingPosition = binaryWriter.Position;
        
        binaryWriter.Position += sizeof(ULONG32);
        
        binaryWriter.Write(isCommitted_);

        ULONG32 endPosition = binaryWriter.Position;
        ULONG32 sizeOfSection = endPosition - startingPosition;

        binaryWriter.Position = startingPosition;
        binaryWriter.Write(sizeOfSection);
        binaryWriter.Position = endPosition;

        replicatedData_ = binaryWriter.GetBuffer(startingPosition);
    }

    operationData.Append(*replicatedData_);
}

ULONG EndTransactionLogRecord::GetSizeOnWire() const
{
    return __super::GetSizeOnWire() + SizeOnWireIncrement;
}

bool EndTransactionLogRecord::Test_Equals(__in LogRecord const & other) const
{
    EndTransactionLogRecord const & otherEndTxLogRecord = dynamic_cast<EndTransactionLogRecord const &>(other);

    if (__super::Test_Equals(other))
    {
        return
            isCommitted_ == otherEndTxLogRecord.isCommitted_;
    }

    return false;
}
